lib/config.js

```diff
@@ -212,127 +87,277 @@ class Config {
+    }
 
-        // Step 4: Merge in command line config file
-        if (this.useSpecificConfig) {
-            debug("Merging command line config file");
+    /**
+     * Gets a list of config objects extracted from local config files that apply to the current directory, in
+     * descending order, beginning with the config that is highest in the directory tree.
+     * @param {string} directory The directory to start looking in for local config files.
+     * @returns {Object[]} The shallow local config objects, or an empty array if there are no local configs.
+     * @private
+     */
+    getLocalConfigHierarchy(directory) {
+        const localConfigFiles = this.findLocalConfigFiles(directory),
+            projectConfigPath = ConfigFile.getFilenameForDirectory(this.options.cwd),
+            searched = [],
+            configs = [];
+        let rootPath,
+            cache;
+
+        for (const localConfigFile of localConfigFiles) {
+            const localConfigDirectory = path.dirname(localConfigFile);
+
+            cache = configCache.getHierarchyLocalConfigs(localConfigDirectory);
+            if (cache) {
+                break;
+            }
 
-            config = ConfigOps.merge(config, this.useSpecificConfig);
-        }
+            // Don't consider the personal config file in the home directory,
+            // except if the home directory is the same as the current working directory
+            if (localConfigDirectory === PERSONAL_CONFIG_DIR && localConfigFile !== projectConfigPath) {
+                continue;
+            }
 
-        // Step 5: Merge in command line environments
-        debug("Merging command line environment settings");
-        config = ConfigOps.merge(config, { env: this.env });
+            // If root flag is set, don't consider file if it is above root
+            if (rootPath && !pathIsInside(path.dirname(localConfigFile), rootPath)) {
+                continue;
+            }
 
-        // Step 6: Merge in command line rules
-        if (this.options.rules) {
-            debug("Merging command line rules");
-            config = ConfigOps.merge(config, { rules: this.options.rules });
-        }
+            debug(`Loading ${localConfigFile}`);
+            const localConfig = ConfigFile.loadCached(localConfigFile);
 
-        // Step 7: Merge in command line globals
-        config = ConfigOps.merge(config, { globals: this.globals });
+            // Ignore empty config files
+            if (!localConfig) {
+                continue;
+            }
 
-        // Only override parser if it is passed explicitly through the command line or if it's not
-        // defined yet (because the final object will at least have the parser key)
-        if (this.parser || !config.parser) {
-            config = ConfigOps.merge(config, {
-                parser: this.parser
-            });
-        }
+            // Check for root flag
+            if (localConfig.root === true) {
```

Comment:
Nitpick: This can just be `if (localConfig.root)` because the property is required to be a boolean if it appears at all.